File: dataforge-storage/src/main/java/hep/dataforge/storage/filestorage/FileEnvelopeLoader.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.storage.filestorage;

import hep.dataforge.exceptions.StorageException;
import hep.dataforge.io.envelopes.DefaultEnvelopeReader;
import hep.dataforge.io.envelopes.DefaultEnvelopeWriter;
import hep.dataforge.io.envelopes.Envelope;
import hep.dataforge.meta.Meta;
import hep.dataforge.storage.api.Storage;
import hep.dataforge.storage.commons.AbstractEnvelopeLoader;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Iterator;

/**
 *
 * @author Alexander Nozik
 */
public class FileEnvelopeLoader extends AbstractEnvelopeLoader {

//    public static FileEnvelopeLoader fromFile(Storage storage, FileObject file, boolean readOnly) throws Exception {
//        try (FileEnvelope envelope = new FileEnvelope(file.getURL().toString(), readOnly)) {
//            if (FileStorageEnvelopeType.validate(envelope)) {
//                FileEnvelopeLoader res = new FileEnvelopeLoader(storage,
//                        FilenameUtils.getBaseName(file.getName().getBaseName()),
//                        envelope.meta(),
//                        file.getURL().toString());
//                res.setReadOnly(readOnly);
//                return res;
//            } else {
//                throw new StorageException("Is not a valid point loader file");
//            }
//        }
//    }

    private final String filePath;
    private FileEnvelope file;

    public FileEnvelopeLoader(Storage storage, String name, Meta meta, String uri) {
        super(storage, name, meta);
        this.filePath = uri;
    }

    private FileEnvelope getFile() throws Exception {
        if (file == null) {
            open();
        }
        return file;
    }

    @Override
    public boolean isEmpty() {
        try {
            return !getFile().hasData();
        } catch (Exception ex) {
            throw new RuntimeException("Can't access loader envelope", ex);
        }
    }

    @Override
    public void open() throws Exception {
        if (this.meta == null) {
            this.meta = getFile().meta();
        }
        if (!isOpen()) {
            file = FileEnvelope.open(filePath, isReadOnly());
        }
    }

    @Override
    public boolean isOpen() {
        return file != null;
    }

    @Override
    public void close() throws Exception {
        getFile().close();
        file = null;
        super.close();
    }

    @Override
    public void push(Envelope env) throws StorageException {
        checkOpen();
        if (!isReadOnly()) {
            try {
                ByteArrayOutputStream baos = new ByteArrayOutputStream();
                DefaultEnvelopeWriter.instance.write(baos, env);
                file.append(baos.toByteArray());
            } catch (IOException ex) {
                throw new StorageException("Can't push envelope to loader", ex);
            }
        } else {
            throw new StorageException("The loader is read only");
        }
    }

    @Override
    public Iterator<Envelope> iterator() {
        checkOpen();
        InputStream st;
        try {
            st = file.getData().getStream();
        } catch (IOException ex) {
            throw new RuntimeException(ex);
        }
        return new Iterator<Envelope>() {
            @Override
            public boolean hasNext() {
                try {
                    return st.available() > 0;
                } catch (IOException ex) {
                    throw new RuntimeException(ex);
                }
            }

            @Override
            public Envelope next() {
                try {
                    return DefaultEnvelopeReader.INSTANCE.readWithData(st);
                } catch (IOException ex) {
                    throw new RuntimeException(ex);
                }
            }
        };
    }

}

File: dataforge-storage/src/main/kotlin/hep/dataforge/storage/filestorage/FileEnvelopeLoader.kt
/*
 * Copyright  2018 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.storage.filestorage

import hep.dataforge.exceptions.StorageException
import hep.dataforge.io.envelopes.DefaultEnvelopeReader
import hep.dataforge.io.envelopes.DefaultEnvelopeWriter
import hep.dataforge.io.envelopes.Envelope
import hep.dataforge.io.jsonMetaType
import hep.dataforge.meta.Meta
import hep.dataforge.storage.api.EnvelopeLoader
import hep.dataforge.storage.api.Storage
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream

/**
 *
 * @author Alexander Nozik
 */
class FileEnvelopeLoader(storage: Storage, name: String, meta: Meta, file: FileEnvelope) : FileLoader(storage, name, meta, file), EnvelopeLoader {

    @Throws(StorageException::class)
    override fun push(env: Envelope) {
        if (!isReadOnly) {
            try {
                val baos = ByteArrayOutputStream()
                DefaultEnvelopeWriter(FileStorageEnvelopeType, jsonMetaType).write(baos, env)
                file.append(baos.toByteArray())
            } catch (ex: IOException) {
                throw StorageException("Can't push envelope to loader", ex)
            }

        } else {
            throw StorageException("The loader is read only")
        }
    }

    override fun iterator(): Iterator<Envelope> {
        val st: InputStream
        try {
            st = file.data.stream
        } catch (ex: IOException) {
            throw RuntimeException(ex)
        }

        return object : Iterator<Envelope> {
            override fun hasNext(): Boolean {
                try {
                    return st.available() > 0
                } catch (ex: IOException) {
                    throw RuntimeException(ex)
                }

            }

            override fun next(): Envelope {
                try {
                    return DefaultEnvelopeReader.INSTANCE.readWithData(st)
                } catch (ex: IOException) {
                    throw RuntimeException(ex)
                }

            }
        }
    }

}
